<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Bringing-in-the-Saved-Dataset-Files" data-toc-modified-id="Bringing-in-the-Saved-Dataset-Files-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Bringing in the Saved Dataset Files</a></span></li><li><span><a href="#Modeling" data-toc-modified-id="Modeling-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Modeling</a></span><ul class="toc-item"><li><span><a href="#Setting-up-Pipelines" data-toc-modified-id="Setting-up-Pipelines-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Setting up Pipelines</a></span></li><li><span><a href="#Setting-Predictor-and-Target-Variables" data-toc-modified-id="Setting-Predictor-and-Target-Variables-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Setting Predictor and Target Variables</a></span><ul class="toc-item"><li><span><a href="#Making-a-Custom-Scorer-to-Cross-Validate-Data" data-toc-modified-id="Making-a-Custom-Scorer-to-Cross-Validate-Data-2.2.1"><span class="toc-item-num">2.2.1&nbsp;&nbsp;</span>Making a Custom Scorer to Cross Validate Data</a></span></li></ul></li><li><span><a href="#Dummy-Model" data-toc-modified-id="Dummy-Model-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Dummy Model</a></span></li><li><span><a href="#Logistic-Regression-Model" data-toc-modified-id="Logistic-Regression-Model-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Logistic Regression Model</a></span><ul class="toc-item"><li><span><a href="#First-Logistic-Regression-GridSearch" data-toc-modified-id="First-Logistic-Regression-GridSearch-2.4.1"><span class="toc-item-num">2.4.1&nbsp;&nbsp;</span>First Logistic Regression GridSearch</a></span></li><li><span><a href="#Second-Logistic-Regression-GridSearch" data-toc-modified-id="Second-Logistic-Regression-GridSearch-2.4.2"><span class="toc-item-num">2.4.2&nbsp;&nbsp;</span>Second Logistic Regression GridSearch</a></span></li><li><span><a href="#Final-Logistic-Regression-GridSearch" data-toc-modified-id="Final-Logistic-Regression-GridSearch-2.4.3"><span class="toc-item-num">2.4.3&nbsp;&nbsp;</span>Final Logistic Regression GridSearch</a></span></li><li><span><a href="#Best-Logistic-Regression-Model" data-toc-modified-id="Best-Logistic-Regression-Model-2.4.4"><span class="toc-item-num">2.4.4&nbsp;&nbsp;</span>Best Logistic Regression Model</a></span></li></ul></li><li><span><a href="#Random-Forest-Model" data-toc-modified-id="Random-Forest-Model-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>Random Forest Model</a></span><ul class="toc-item"><li><span><a href="#First-Random-Forest-GridSearch" data-toc-modified-id="First-Random-Forest-GridSearch-2.5.1"><span class="toc-item-num">2.5.1&nbsp;&nbsp;</span>First Random Forest GridSearch</a></span></li><li><span><a href="#Second-Random-Forest-Gridsearch" data-toc-modified-id="Second-Random-Forest-Gridsearch-2.5.2"><span class="toc-item-num">2.5.2&nbsp;&nbsp;</span>Second Random Forest Gridsearch</a></span></li><li><span><a href="#Third-Random-Forest-Gridsearch" data-toc-modified-id="Third-Random-Forest-Gridsearch-2.5.3"><span class="toc-item-num">2.5.3&nbsp;&nbsp;</span>Third Random Forest Gridsearch</a></span></li><li><span><a href="#Best-Random-Forest-Model" data-toc-modified-id="Best-Random-Forest-Model-2.5.4"><span class="toc-item-num">2.5.4&nbsp;&nbsp;</span>Best Random Forest Model</a></span></li></ul></li><li><span><a href="#XGBoost-Model" data-toc-modified-id="XGBoost-Model-2.6"><span class="toc-item-num">2.6&nbsp;&nbsp;</span>XGBoost Model</a></span><ul class="toc-item"><li><span><a href="#XGBoost-Gridsearch" data-toc-modified-id="XGBoost-Gridsearch-2.6.1"><span class="toc-item-num">2.6.1&nbsp;&nbsp;</span>XGBoost Gridsearch</a></span></li><li><span><a href="#Best-XGBoost-Model" data-toc-modified-id="Best-XGBoost-Model-2.6.2"><span class="toc-item-num">2.6.2&nbsp;&nbsp;</span>Best XGBoost Model</a></span></li></ul></li><li><span><a href="#LinearSVC-Model" data-toc-modified-id="LinearSVC-Model-2.7"><span class="toc-item-num">2.7&nbsp;&nbsp;</span>LinearSVC Model</a></span><ul class="toc-item"><li><span><a href="#SVC-GridSearch" data-toc-modified-id="SVC-GridSearch-2.7.1"><span class="toc-item-num">2.7.1&nbsp;&nbsp;</span>SVC GridSearch</a></span></li></ul></li><li><span><a href="#MLP-Model" data-toc-modified-id="MLP-Model-2.8"><span class="toc-item-num">2.8&nbsp;&nbsp;</span>MLP Model</a></span><ul class="toc-item"><li><span><a href="#First-MLPClassifier-Gridsearch" data-toc-modified-id="First-MLPClassifier-Gridsearch-2.8.1"><span class="toc-item-num">2.8.1&nbsp;&nbsp;</span>First MLPClassifier Gridsearch</a></span></li><li><span><a href="#Second-MLPClassifier-GridSearch" data-toc-modified-id="Second-MLPClassifier-GridSearch-2.8.2"><span class="toc-item-num">2.8.2&nbsp;&nbsp;</span>Second MLPClassifier GridSearch</a></span></li></ul></li><li><span><a href="#K-Nearest-Neighbors-Model" data-toc-modified-id="K-Nearest-Neighbors-Model-2.9"><span class="toc-item-num">2.9&nbsp;&nbsp;</span>K-Nearest Neighbors Model</a></span><ul class="toc-item"><li><span><a href="#First-K-Nearest-Neighbors-Gridsearch" data-toc-modified-id="First-K-Nearest-Neighbors-Gridsearch-2.9.1"><span class="toc-item-num">2.9.1&nbsp;&nbsp;</span>First K-Nearest Neighbors Gridsearch</a></span></li><li><span><a href="#Second-K-Nearest-Neighbors-GridSearch" data-toc-modified-id="Second-K-Nearest-Neighbors-GridSearch-2.9.2"><span class="toc-item-num">2.9.2&nbsp;&nbsp;</span>Second K-Nearest Neighbors GridSearch</a></span></li><li><span><a href="#K-Nearest-Neighbors-Best-Model" data-toc-modified-id="K-Nearest-Neighbors-Best-Model-2.9.3"><span class="toc-item-num">2.9.3&nbsp;&nbsp;</span>K-Nearest Neighbors Best Model</a></span></li></ul></li><li><span><a href="#Stacking-Classifier-Model" data-toc-modified-id="Stacking-Classifier-Model-2.10"><span class="toc-item-num">2.10&nbsp;&nbsp;</span>Stacking Classifier Model</a></span><ul class="toc-item"><li><span><a href="#Second-Stacking-Classifier-Model" data-toc-modified-id="Second-Stacking-Classifier-Model-2.10.1"><span class="toc-item-num">2.10.1&nbsp;&nbsp;</span>Second Stacking Classifier Model</a></span></li><li><span><a href="#Third-Iteration-of-Stacking-Classifier" data-toc-modified-id="Third-Iteration-of-Stacking-Classifier-2.10.2"><span class="toc-item-num">2.10.2&nbsp;&nbsp;</span>Third Iteration of Stacking Classifier</a></span></li><li><span><a href="#Final-Iteration-of-Stacking-Classifier" data-toc-modified-id="Final-Iteration-of-Stacking-Classifier-2.10.3"><span class="toc-item-num">2.10.3&nbsp;&nbsp;</span>Final Iteration of Stacking Classifier</a></span></li></ul></li></ul></li><li><span><a href="#Best-Model" data-toc-modified-id="Best-Model-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Best Model</a></span><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#Visualizing-Errors" data-toc-modified-id="Visualizing-Errors-3.0.1"><span class="toc-item-num">3.0.1&nbsp;&nbsp;</span>Visualizing Errors</a></span></li></ul></li></ul></li><li><span><a href="#Conclusions" data-toc-modified-id="Conclusions-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Conclusions</a></span><ul class="toc-item"><li><span><a href="#Analysis" data-toc-modified-id="Analysis-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Analysis</a></span></li><li><span><a href="#Future-Considerations" data-toc-modified-id="Future-Considerations-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Future Considerations</a></span></li></ul></li></ul></div>

Running main notebook

In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np

import geopandas as gpd
from shapely.geometry import Point
import geopy.distance as distance

import planetary_computer as pc
from pystac_client import Client

from datetime import datetime
from datetime import timedelta

# import requests
# from PIL import Image
# from io import BytesIO

# import rioxarray
# import odc.stac
# import tempfile
# import os

from sklearn.preprocessing import MinMaxScaler, OneHotEncoder, StandardScaler
from sklearn.model_selection import train_test_split, cross_validate, GridSearchCV
from sklearn.ensemble import RandomForestClassifier
from sklearn.dummy import DummyClassifier
from sklearn.pipeline import Pipeline
from sklearn.impute import KNNImputer
from sklearn.compose import ColumnTransformer
from sklearn.metrics import classification_report, mean_squared_error, make_scorer
from sklearn.neural_network import MLPClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.ensemble import StackingClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.svm import LinearSVC

from xgboost import XGBClassifier

import hab_functions

# import pygrib

import warnings
warnings.filterwarnings('ignore')

%matplotlib inline
sns.set_theme()

# Bringing in the Saved Dataset Files

For the modeling portion of this report, I will be using the full dataset that has undergone the same cleaning that we see in the `Introduction_and_Data` notebook. It includes NASA Satellite Imagery, elevation data, and NOAA Climate data in the form of temperature for one and five day's before the manual sample was taken.

In [2]:
full_train = pd.read_csv('./Data/clean_train_data.csv')

In [3]:
full_train.drop('Unnamed: 0', axis=1, inplace=True)
full_train.head()

date   uid   region   latitude  longitude  img_date    red_mean  \
0  734872  evep  midwest  44.847993 -93.476318    734872    0.000000   
1  734872  paev  midwest  44.822478 -93.367962    734872    0.000000   
2  734872  gdxr  midwest  44.877646 -93.557842    734863  178.941946   
3  734872  guny  midwest  44.878889 -93.490833    734872    0.000000   
4  734872  fwbt  midwest  44.850500 -93.515700    734872    0.000000   

   red_median  red_max  red_min  ...  severity  missing_img_date  \
0    0.000000      0.0      0.0  ...         1                 0   
1    0.000000      0.0      0.0  ...         1                 0   
2  204.755425    255.0      0.0  ...         1                 0   
3    0.000000      0.0      0.0  ...         1                 0   
4    0.000000      0.0      0.0  ...         1                 0   

   days_from_sat_to_sample  season_autumn  season_spring  season_summer  \
0                        0            0.0            0.0            0.0   
1                        0            0.0            0.0            0.0   
2                        9            0.0            0.0            0.0   
3                        0            0.0            0.0            0.0   
4                        0            0.0            0.0            0.0   

   season_winter   elevation  temp_1day  temp_5_day  
0            1.0  254.000000        NaN         NaN  
1            1.0  247.543610        NaN         NaN  
2            1.0  289.865173        NaN         NaN  
3            1.0  277.500000        NaN         NaN  
4            1.0  265.504028        NaN         NaN  

[5 rows x 34 columns]

In [4]:
full_test_df = pd.read_csv('./Data/clean_test_data.csv')

In [5]:
full_test_df.drop('Unnamed: 0', axis=1, inplace=True)
full_test_df.head()

date   uid  latitude  longitude  img_date    red_mean  red_median  \
0  734876  eamn   36.9818   -120.221    734872    0.000000    0.000000   
1  734876  howw   37.0062   -120.600    734863   96.794956   93.012939   
2  734876  imsv   36.9836   -120.500    734863  165.295561  178.719942   
3  734893  rsos   33.8892   -117.562    734890   83.199385   73.647525   
4  734893  wgxq   33.8011   -117.205    734890  108.140408  100.092696   

   red_max  red_min      red_sum  ...  blue_product  missing_img_date  \
0      0.0      0.0     0.000000  ...           0.0                 0   
1    255.0      0.0  5420.517560  ...           0.0                 0   
2    255.0      0.0  8099.482488  ...           0.0                 0   
3    255.0      0.0  4076.769845  ...           0.0                 0   
4    255.0      0.0  6920.986096  ...           0.0                 0   

   days_from_sat_to_sample  season_autumn  season_spring  season_summer  \
0                        4            0.0            0.0            0.0   
1                       13            0.0            0.0            0.0   
2                       13            0.0            0.0            0.0   
3                        3            0.0            0.0            0.0   
4                        3            0.0            0.0            0.0   

   season_winter   elevation  temp_1day  temp_5_day  
0            1.0   62.286789        NaN         NaN  
1            1.0   33.658421        NaN         NaN  
2            1.0   36.339100        NaN         NaN  
3            1.0  182.744644        NaN         NaN  
4            1.0  433.029602        NaN         NaN  

[5 rows x 32 columns]

# Modeling

## Setting up Pipelines

Creating a pipeline to fill in missing values and standard scale some features. Imputing and scaling apply to sample site coordinates, to all features brought in from Satellites including images and elevation, and to temperature data from NOAA.

I'm utilizing a K-Nearest Neighbor approach for imputing data. Any data that is being imputed will have a missing indicator added so that the model can be aware of imputation.

In [6]:
impute_scale_pipe = Pipeline([
    ('knn_impute', KNNImputer(add_indicator=True)),
    ('ss', StandardScaler())
])

In [7]:
# with lat/long
impute_scale_cols = ['latitude', 'longitude', 'red_mean', 'red_median', 'red_max', 'red_min','red_sum',
        'red_product', 'green_mean', 'green_median', 'green_max',
        'green_min', 'green_sum', 'green_product', 'blue_mean',
        'blue_median','blue_max', 'blue_min', 'blue_sum', 'blue_product',
        'elevation', 'temp_1day', 'temp_5_day']

ColTrans = ColumnTransformer(transformers=[
    ('impute_scale_pipe', impute_scale_pipe, impute_scale_cols)
    
], remainder='passthrough')

## Setting Predictor and Target Variables

Here I'll be creating the predictor and target variables as well as completing a train/test split to create validation data. I will be using the same ratio for the split as we have on the train/test data.

In [8]:
X = full_train.drop(['severity','uid', 'region'], axis=1)
y = full_train['severity']
score = full_train[['uid', 'region']]

In [9]:
full_test = full_test_df.drop('uid', axis=1)
full_test_uid = full_test_df[['uid']]

In [10]:
# setting the val size to the same percent as the real train/test split
X_train, X_val, y_train, y_val = train_test_split(X, y, test_size=.276, random_state=42)

### Making a Custom Scorer to Cross Validate Data

To score the results of the models, I will be using a custom scorer. It takes the regional Root Mean Squared Error and averages them. I use this approach because of the different levels of sampling in each region to give them each equal weight.

In [11]:
val_score = score.join(y_val, how='right')

In [12]:
val_score

uid     region  severity
13699  vmps      south         3
3143   nhqm      south         1
12222  anmy    midwest         1
13666  ltvs       west         5
720    pljh  northeast         3
...     ...        ...       ...
12546  jiqv    midwest         2
7093   fbdi      south         1
16934  apuu    midwest         3
6332   lovo    midwest         2
6608   khum    midwest         4

[4709 rows x 3 columns]

In [13]:
# This function lets me spot check a score

def check_score(val_score, val_pred):
    val_score['pred'] = val_pred
    region_scores = []
    for region in val_score.region.unique():
        sub = val_score[val_score.region == region]
        region_rmse = mean_squared_error(sub.severity, sub.pred, squared=False)
        print(f"RMSE for {region} (n={len(sub)}): {round(region_rmse, 4)}")
        region_scores.append(region_rmse)

    overall_rmse = np.mean(region_scores)
    print(f"Final score: {overall_rmse}")
    return overall_rmse

In [14]:
# this function is able to be turned into a custom scorer for cv and gridsearch

def custom_scorer(y_val, val_pred):
    val_score = score.join(y_val, how='right')
    val_score['pred'] = val_pred
    region_scores = []
    for region in val_score.region.unique():
        sub = val_score[val_score.region == region]
        region_rmse = mean_squared_error(sub.severity, sub.pred, squared=False)
        region_scores.append(region_rmse)

    overall_rmse = np.mean(region_scores)
    return overall_rmse

In [15]:
regional_scorer = make_scorer(custom_scorer, greater_is_better=False)

In [93]:
scores_df = pd.DataFrame()

# Function to update scorecard as models run
def update_scorecard(model, model_name, scores_df):
    # score on unseen data
    if ('oost') or ('ack') in str(model):
        model.fit(X_train, y_train_adj);
        pred = model.predict(X_val)
        pred=pred+1
        score = check_score(val_score, pred)

        # cv scores
        cv = cross_validate(model, 
                            X_train, 
                            y_train_adj, 
                            scoring=regional_scorer, 
                            return_train_score=True)
    else:
        model.fit(X_train, y_train);
        pred = model.predict(X_val)
        score = check_score(val_score, pred)

        # cv scores
        cv = cross_validate(model, 
                            X_train, 
                            y_train, 
                            scoring=regional_scorer, 
                            return_train_score=True)
    
    # update scorecard
    # creating a dictionary to store results
    score_dict = {'Model Name':model_name}
    # appending to dict and df
    score_dict['CV_Train'] = abs(cv['train_score'].mean())
    score_dict['CV_Test'] = abs(cv['test_score'].mean())
    difference = score_dict['CV_Test'] - score_dict['CV_Train']
    difference = np.round(difference, 3)
    if difference >0:
        score_dict['Difference'] = f'Overfit: ({abs(difference)})'
    elif difference <0:
        score_dict['Difference'] = f'Underfit: ({abs(difference)})'
    else:
        score_dict['Difference'] = f'Good Fit: ({abs(difference)})'
    score_dict['Unseen Data'] = score
    scores_df = scores_df.append(score_dict, ignore_index=True)
    return(scores_df)

## Dummy Model

My Dummy Model performs as expected when predicting the most frequent severity level for each sample point. The final Regional RMSE score on unseen validation data for this model is 1.65.

In [17]:
dummy_model = Pipeline([
    ('ColTrans', ColTrans),
    ('dummy', DummyClassifier(strategy='most_frequent', random_state=42))
])

In [18]:
dummy_model.fit(X_train, y_train);

In [19]:
y_val.value_counts(normalize=True)

1    0.433213
4    0.209174
2    0.193247
3    0.161605
5    0.002761
Name: severity, dtype: float64

In [20]:
dummy_model.score(X_val, y_val)

0.4332129963898917

In [21]:
dummy_val_pred = dummy_model.predict(X_val)

In [22]:
dummy_score = check_score(val_score, dummy_val_pred)

RMSE for south (n=2744): 0.9751
RMSE for midwest (n=626): 1.5982
RMSE for west (n=1030): 2.8527
RMSE for northeast (n=309): 1.1892
Final score: 1.6538061708774303


Updating the scorecard.

In [23]:
scores_df = update_scorecard(dummy_model, 'dummy_model', scores_df)
scores_df

RMSE for south (n=2744): 0.9751
RMSE for midwest (n=626): 1.5982
RMSE for west (n=1030): 2.8527
RMSE for northeast (n=309): 1.1892
Final score: 1.6538061708774303


Model Name  CV_Train   CV_Test         Difference  Unseen Data
0  dummy_model  1.658196  1.657582  Underfit: (0.001)     1.653806

## Logistic Regression Model

The baseline Logistic Regression Model performs exactly the same as the dummy model.

In [24]:
log_reg_model = Pipeline([
    ('ColTrans', ColTrans),
    ('logreg', LogisticRegression(random_state=42))
])

In [25]:
log_reg_model.fit(X_train, y_train);

In [26]:
log_reg_model.score(X_val, y_val)

0.4332129963898917

In [27]:
log_reg_pred = log_reg_model.predict(X_val)

In [28]:
lr_base_score = check_score(val_score, log_reg_pred)

RMSE for south (n=2744): 0.9751
RMSE for midwest (n=626): 1.5982
RMSE for west (n=1030): 2.8527
RMSE for northeast (n=309): 1.1892
Final score: 1.6538061708774303


In [29]:
log_reg_cv = cross_validate(log_reg_model, 
                            X_train, 
                            y_train, 
                            scoring=regional_scorer, 
                            return_train_score=True)

In [30]:
print(f"Train: {log_reg_cv['train_score'].mean()}")
print(f"Test: {log_reg_cv['test_score'].mean()}")
log_reg_cv

Train: -1.658196064281532
Test: -1.6575819051725245


{'fit_time': array([1.22833014, 1.3437047 , 1.23503089, 1.21620798, 1.27515841]),
 'score_time': array([0.31346989, 0.30414319, 0.25000978, 0.31065917, 0.30948281]),
 'test_score': array([-1.64969162, -1.62809574, -1.66571971, -1.65495332, -1.68944913]),
 'train_score': array([-1.65997003, -1.66581971, -1.65633035, -1.65907389, -1.64978634])}

Updating the scorecard.

In [31]:
scores_df = update_scorecard(log_reg_model, 'LogReg_Base', scores_df)
scores_df

RMSE for south (n=2744): 0.9751
RMSE for midwest (n=626): 1.5982
RMSE for west (n=1030): 2.8527
RMSE for northeast (n=309): 1.1892
Final score: 1.6538061708774303


Model Name  CV_Train   CV_Test         Difference  Unseen Data
0  dummy_model  1.658196  1.657582  Underfit: (0.001)     1.653806
1  LogReg_Base  1.658196  1.657582  Underfit: (0.001)     1.653806

### First Logistic Regression GridSearch

The first implementation of GridSearch greatly increased the performance of the model, going from a score of 1.65 on the validation data to a cross-validated score of .984.

In [32]:
log_params = {
    'logreg__solver':['liblinear', 'sag', 'saga', 'newton-cg', 'lbfgs'],
    'logreg__fit_intercept':[True, False]
}

In [33]:
log_reg_gs = GridSearchCV(estimator=log_reg_model, 
                          param_grid=log_params,
                          scoring=regional_scorer,
                          cv=5, 
                          verbose=1,
                          n_jobs=2)

log_reg_gs.fit(X_train, y_train);

Fitting 5 folds for each of 10 candidates, totalling 50 fits


In [34]:
log_reg_gs.best_params_

{'logreg__fit_intercept': False, 'logreg__solver': 'newton-cg'}

In [35]:
log_reg_gs.best_score_

-0.9841963009173653

Updating the scorecard.

In [36]:
scores_df = update_scorecard(log_reg_gs.best_estimator_, 'LogReg_GS1', scores_df)
scores_df

RMSE for south (n=2744): 0.9683
RMSE for midwest (n=626): 1.1269
RMSE for west (n=1030): 0.526
RMSE for northeast (n=309): 1.1892
Final score: 0.9526271057193282


Model Name  CV_Train   CV_Test         Difference  Unseen Data
0  dummy_model  1.658196  1.657582  Underfit: (0.001)     1.653806
1  LogReg_Base  1.658196  1.657582  Underfit: (0.001)     1.653806
2   LogReg_GS1  0.983089  0.984196   Overfit: (0.001)     0.952627

### Second Logistic Regression GridSearch

The second iteration of GridSearch continued to see improvement on the final cross validated results. Moving from .984 down to .982. I will complete one final iteration of GridSearch to find the best model.

In [37]:
log_params2 = {
    'logreg__solver':['saga', 'newton-cg', 'lbfgs'],
    'logreg__fit_intercept':[True, False],
    'logreg__max_iter':[100, 150, 200]
}

In [38]:
log_reg_gs2 = GridSearchCV(estimator=log_reg_model, 
                          param_grid=log_params2,
                          scoring=regional_scorer,
                          cv=5, 
                          verbose=1,
                          n_jobs=2)

log_reg_gs2.fit(X_train, y_train);

Fitting 5 folds for each of 18 candidates, totalling 90 fits


In [39]:
log_reg_gs2.best_params_

{'logreg__fit_intercept': False,
 'logreg__max_iter': 200,
 'logreg__solver': 'newton-cg'}

In [40]:
log_reg_gs2.best_score_

-0.9820592442530044

Updating the scorecard.

In [41]:
scores_df = update_scorecard(log_reg_gs2.best_estimator_, 'LogReg_GS2', scores_df)
scores_df

RMSE for south (n=2744): 0.9685
RMSE for midwest (n=626): 1.0975
RMSE for west (n=1030): 0.5513
RMSE for northeast (n=309): 1.1892
Final score: 0.9516217910789043


Model Name  CV_Train   CV_Test         Difference  Unseen Data
0  dummy_model  1.658196  1.657582  Underfit: (0.001)     1.653806
1  LogReg_Base  1.658196  1.657582  Underfit: (0.001)     1.653806
2   LogReg_GS1  0.983089  0.984196   Overfit: (0.001)     0.952627
3   LogReg_GS2  0.978392  0.982059   Overfit: (0.004)     0.951622

### Final Logistic Regression GridSearch

The final Iteration of GridSearch identifies the following hyperparameters as the best and returns a final score of .981.

- Fit_intercept = False
- Max_iter = 175
- Penalty = None
- Solver = newton_cg

In [42]:
log_params3 = {
    'logreg__solver':['newton-cg'],
    'logreg__fit_intercept':[False],
    'logreg__max_iter':[175, 200],
    'logreg__penalty':[None, 'l2']
}

In [43]:
log_reg_gs3 = GridSearchCV(estimator=log_reg_model, 
                          param_grid=log_params3,
                          scoring=regional_scorer,
                          cv=5, 
                          verbose=1,
                          n_jobs=2)

log_reg_gs3.fit(X_train, y_train);

Fitting 5 folds for each of 4 candidates, totalling 20 fits


In [44]:
log_reg_gs3.best_params_

{'logreg__fit_intercept': False,
 'logreg__max_iter': 175,
 'logreg__penalty': None,
 'logreg__solver': 'newton-cg'}

In [45]:
log_reg_gs3.best_score_

-0.9811093931102773

Updating the scorecard.

In [46]:
scores_df = update_scorecard(log_reg_gs3.best_estimator_, 'LogReg_GS3', scores_df)
scores_df

RMSE for south (n=2744): 0.9683
RMSE for midwest (n=626): 1.0946
RMSE for west (n=1030): 0.5583
RMSE for northeast (n=309): 1.1892
Final score: 0.9525960995468835


Model Name  CV_Train   CV_Test         Difference  Unseen Data
0  dummy_model  1.658196  1.657582  Underfit: (0.001)     1.653806
1  LogReg_Base  1.658196  1.657582  Underfit: (0.001)     1.653806
2   LogReg_GS1  0.983089  0.984196   Overfit: (0.001)     0.952627
3   LogReg_GS2  0.978392  0.982059   Overfit: (0.004)     0.951622
4   LogReg_GS3  0.974411  0.981109   Overfit: (0.007)     0.952596

### Best Logistic Regression Model

This is a pretty decent model that is only slightly overfit on training data. Cross validated test score: .981 and train score .974.

In [47]:
best_log_reg_model = Pipeline([
    ('ColTrans', ColTrans),
    ('logreg', LogisticRegression(random_state=42,
                                 fit_intercept=False,
                                 max_iter=175,
                                 penalty=None,
                                 solver='newton-cg'))
])

In [48]:
best_log_reg_cv = cross_validate(best_log_reg_model, 
                                 X_train, 
                                 y_train, 
                                 scoring=regional_scorer, 
                                 return_train_score=True,
                                 n_jobs=2)

In [49]:
print(f"Train: {best_log_reg_cv['train_score'].mean()}")
print(f"Test: {best_log_reg_cv['test_score'].mean()}")
best_log_reg_cv

Train: -0.9744108414872275
Test: -0.9811093931102773


{'fit_time': array([7.1336987 , 6.18027854, 6.89961076, 5.62131238, 5.00644636]),
 'score_time': array([0.32253146, 0.3533113 , 0.31552601, 0.36406946, 0.31916285]),
 'test_score': array([-0.97934291, -0.94962886, -0.97520756, -0.99544208, -1.00592554]),
 'train_score': array([-0.97378052, -0.97508696, -0.97036751, -0.97168104, -0.98113818])}

**Testing on Validation Data**

Final score on unseen data is .95.

In [50]:
best_log_reg_model.fit(X_train, y_train);

In [51]:
best_log_reg_pred = best_log_reg_model.predict(X_val)

In [52]:
check_score(val_score, best_log_reg_pred)

RMSE for south (n=2744): 0.9683
RMSE for midwest (n=626): 1.0946
RMSE for west (n=1030): 0.5583
RMSE for northeast (n=309): 1.1892
Final score: 0.9525960995468835


0.9525960995468835

## Random Forest Model

The baseline random forest model performs well with a Regional RMSE score of .836 on cross-validated data. The model is definitely overfit and can hopefully see better performance through hyperparameter tuning and GridSearch.

In [53]:
rf_model = Pipeline([
    ('ColTrans', ColTrans),
    ('rf', RandomForestClassifier(random_state=42))
])

In [54]:
rf_model.fit(X_train, y_train);

In [55]:
rf_model.score(X_val, y_val)

0.6476959014652792

In [56]:
rf_val_pred = rf_model.predict(X_val)

In [57]:
check_score(val_score, rf_val_pred)

RMSE for south (n=2744): 0.9233
RMSE for midwest (n=626): 0.8766
RMSE for west (n=1030): 0.4439
RMSE for northeast (n=309): 0.8495
Final score: 0.7733224741017735


0.7733224741017735

In [58]:
rf_cv = cross_validate(rf_model, 
                       X_train, 
                       y_train, 
                       scoring=regional_scorer, 
                       return_train_score=True, 
                       n_jobs=2)

In [59]:
print(f"Train: {rf_cv['train_score'].mean()}")
print(f"Test: {rf_cv['test_score'].mean()}")
rf_cv

Train: -0.004529944947442045
Test: -0.8363883564253282


{'fit_time': array([3.00968766, 3.18719983, 2.98893237, 2.90740824, 2.85704947]),
 'score_time': array([0.36653113, 0.46071076, 0.30553436, 0.41595244, 0.34179425]),
 'test_score': array([-0.83079048, -0.80881587, -0.79806215, -0.85972863, -0.88454466]),
 'train_score': array([-0.00330698, -0.0086926 , -0.00735867, -0.        , -0.00329147])}

Updating the scorecard.

In [60]:
scores_df = update_scorecard(rf_model, 'Random_Forest_Base', scores_df)
scores_df

RMSE for south (n=2744): 0.9233
RMSE for midwest (n=626): 0.8766
RMSE for west (n=1030): 0.4439
RMSE for northeast (n=309): 0.8495
Final score: 0.7733224741017735


Model Name  CV_Train   CV_Test         Difference  Unseen Data
0         dummy_model  1.658196  1.657582  Underfit: (0.001)     1.653806
1         LogReg_Base  1.658196  1.657582  Underfit: (0.001)     1.653806
2          LogReg_GS1  0.983089  0.984196   Overfit: (0.001)     0.952627
3          LogReg_GS2  0.978392  0.982059   Overfit: (0.004)     0.951622
4          LogReg_GS3  0.974411  0.981109   Overfit: (0.007)     0.952596
5  Random_Forest_Base  0.004530  0.836388   Overfit: (0.832)     0.773322

### First Random Forest GridSearch

Since our data has a class imbalance, we will be testing to see whether class weights help to improve the performance of the model. Additional, we'll look to see if reducing the number of samples helps with model performance. 

The first iteration of GridSearch improved the Regional RMSE score from .836 to .831.

In [61]:
rf_params = {
    'rf__class_weight':[None, 'balanced'],
    'rf__max_samples':[None, .5, .75]
}
rf_gs = GridSearchCV(estimator=rf_model, 
                     param_grid=rf_params, 
                     scoring=regional_scorer, 
                     cv=5, 
                     verbose=1, 
                     return_train_score=True,
                     n_jobs=2)

rf_gs.fit(X_train, y_train);

Fitting 5 folds for each of 6 candidates, totalling 30 fits


In [62]:
rf_gs.best_params_

{'rf__class_weight': None, 'rf__max_samples': 0.75}

In [63]:
rf_gs.best_score_

-0.8316550856065819

Updating the scorecard.

In [64]:
scores_df = update_scorecard(rf_gs.best_estimator_, 'Random_Forest_GS1', scores_df)
scores_df

RMSE for south (n=2744): 0.9284
RMSE for midwest (n=626): 0.8999
RMSE for west (n=1030): 0.4192
RMSE for northeast (n=309): 0.9049
Final score: 0.7880951444841156


Model Name  CV_Train   CV_Test         Difference  Unseen Data
0         dummy_model  1.658196  1.657582  Underfit: (0.001)     1.653806
1         LogReg_Base  1.658196  1.657582  Underfit: (0.001)     1.653806
2          LogReg_GS1  0.983089  0.984196   Overfit: (0.001)     0.952627
3          LogReg_GS2  0.978392  0.982059   Overfit: (0.004)     0.951622
4          LogReg_GS3  0.974411  0.981109   Overfit: (0.007)     0.952596
5  Random_Forest_Base  0.004530  0.836388   Overfit: (0.832)     0.773322
6   Random_Forest_GS1  0.052565  0.831655   Overfit: (0.779)     0.788095

### Second Random Forest Gridsearch

The second GridSearch further improved the score of the model from .831 to .814 on cross validated data. I will attempt one more iteration of GridSearch to find the best model.

In [65]:
rf_params2 = {
    'rf__criterion':['gini', 'entropy', 'log_loss'],
    'rf__min_samples_split':[2, 3, 5],
    'rf__max_samples':[None, .5, .75]
}
rf_gs2 = GridSearchCV(estimator=rf_model, 
                      param_grid=rf_params2, 
                      scoring=regional_scorer, 
                      cv=5, 
                      verbose=1, 
                      return_train_score=True,
                      n_jobs=4)

rf_gs2.fit(X_train, y_train);

Fitting 5 folds for each of 27 candidates, totalling 135 fits


In [66]:
rf_gs2.best_params_

{'rf__criterion': 'entropy',
 'rf__max_samples': None,
 'rf__min_samples_split': 3}

In [67]:
rf_gs2.best_score_

-0.814660181569517

Updating the scorecard.

In [68]:
scores_df = update_scorecard(rf_gs2.best_estimator_, 'Random_Forest_GS2', scores_df)
scores_df

RMSE for south (n=2744): 0.927
RMSE for midwest (n=626): 0.8784
RMSE for west (n=1030): 0.4483
RMSE for northeast (n=309): 0.8476
Final score: 0.7753245739834157


Model Name  CV_Train   CV_Test         Difference  Unseen Data
0         dummy_model  1.658196  1.657582  Underfit: (0.001)     1.653806
1         LogReg_Base  1.658196  1.657582  Underfit: (0.001)     1.653806
2          LogReg_GS1  0.983089  0.984196   Overfit: (0.001)     0.952627
3          LogReg_GS2  0.978392  0.982059   Overfit: (0.004)     0.951622
4          LogReg_GS3  0.974411  0.981109   Overfit: (0.007)     0.952596
5  Random_Forest_Base  0.004530  0.836388   Overfit: (0.832)     0.773322
6   Random_Forest_GS1  0.052565  0.831655   Overfit: (0.779)     0.788095
7   Random_Forest_GS2  0.076151  0.814660   Overfit: (0.739)     0.775325

### Third Random Forest Gridsearch

The third GridSearch continued to see improved results bringing the regional score from .814 to .808.

In [69]:
rf_params3 = {
    'rf__criterion':['gini', 'entropy'],
    'rf__min_samples_split':[3, 4, 6],
    'rf__n_estimators':[150, 200, 300]
}
rf_gs3 = GridSearchCV(estimator=rf_model, 
                      param_grid=rf_params3, 
                      scoring=regional_scorer, 
                      cv=5, 
                      verbose=1, 
                      n_jobs=4)

rf_gs3.fit(X_train, y_train);

Fitting 5 folds for each of 18 candidates, totalling 90 fits


In [70]:
rf_gs3.best_params_

{'rf__criterion': 'entropy',
 'rf__min_samples_split': 3,
 'rf__n_estimators': 150}

In [71]:
rf_gs3.best_score_

-0.8081060998387551

Updating the scorecard.

In [72]:
scores_df = update_scorecard(rf_gs3.best_estimator_, 'Random_Forest_GS3', scores_df)
scores_df

RMSE for south (n=2744): 0.9246
RMSE for midwest (n=626): 0.8811
RMSE for west (n=1030): 0.4272
RMSE for northeast (n=309): 0.9013
Final score: 0.7835635905214304


Model Name  CV_Train   CV_Test         Difference  Unseen Data
0         dummy_model  1.658196  1.657582  Underfit: (0.001)     1.653806
1         LogReg_Base  1.658196  1.657582  Underfit: (0.001)     1.653806
2          LogReg_GS1  0.983089  0.984196   Overfit: (0.001)     0.952627
3          LogReg_GS2  0.978392  0.982059   Overfit: (0.004)     0.951622
4          LogReg_GS3  0.974411  0.981109   Overfit: (0.007)     0.952596
5  Random_Forest_Base  0.004530  0.836388   Overfit: (0.832)     0.773322
6   Random_Forest_GS1  0.052565  0.831655   Overfit: (0.779)     0.788095
7   Random_Forest_GS2  0.076151  0.814660   Overfit: (0.739)     0.775325
8   Random_Forest_GS3  0.067371  0.808106   Overfit: (0.741)     0.783564

### Best Random Forest Model

The best model is still very overfit but produces a cross validated score of .808 for test data and .067 on training data. Final Score on the unseen validation data is .783.

In [73]:
best_rf_model = Pipeline([
    ('ColTrans', ColTrans),
    ('rf', RandomForestClassifier(criterion='entropy',
                                  min_samples_split=3,
                                  n_estimators=150,
                                  random_state=42))
])

In [74]:
best_rf_cv = cross_validate(best_rf_model, 
                            X_train, 
                            y_train, 
                            scoring=regional_scorer, 
                            return_train_score=True,
                            n_jobs=2)

In [75]:
print(f"Train: {best_rf_cv['train_score'].mean()}")
print(f"Test: {best_rf_cv['test_score'].mean()}")
best_rf_cv

Train: -0.06737069080285488
Test: -0.8081060998387551


{'fit_time': array([4.80354786, 4.91306567, 4.74617338, 4.70239949, 4.6266737 ]),
 'score_time': array([0.39706731, 0.45060372, 0.31678605, 0.44407058, 0.36621547]),
 'test_score': array([-0.79269323, -0.80051718, -0.77944494, -0.82543635, -0.84243879]),
 'train_score': array([-0.07606819, -0.07727563, -0.0620053 , -0.06328816, -0.05821616])}

**Testing on Validation Data**

In [76]:
best_rf_model.fit(X_train, y_train);

In [77]:
best_rf_val_pred = best_rf_model.predict(X_val)

In [78]:
check_score(val_score, best_rf_val_pred)

RMSE for south (n=2744): 0.9246
RMSE for midwest (n=626): 0.8811
RMSE for west (n=1030): 0.4272
RMSE for northeast (n=309): 0.9013
Final score: 0.7835635905214304


0.7835635905214304

## XGBoost Model

XGBoost requires that classes begin with zero, I have created an adjusted y variable to account for the requirement. 

XGBoost has a great score to begin with, unseen validation data is scoring in at .732 and the cross validation on training data is scoring at .768. The model is overfit though with training score coming in at .271 and could see some improvement from cross validating.

In [79]:
y_adjusted = y.apply(lambda x: x-1)

In [80]:
# setting the val size to the same percent as the real train/test split
X_train, X_val, y_train_adj, y_val_adj = train_test_split(X, y_adjusted, test_size=.276, random_state=42)

In [81]:
xgb_model = Pipeline([
    ('ColTrans', ColTrans),
    ('xgb', XGBClassifier(random_state=42))
])

In [82]:
xgb_model.fit(X_train, y_train_adj);

In [83]:
xgb_model.score(X_val, y_val_adj)

0.6689318326608622

In [84]:
xgb_val_pred = xgb_model.predict(X_val)
xgb_val_pred = xgb_val_pred+1

In [85]:
check_score(val_score, xgb_val_pred)

RMSE for south (n=2744): 0.9031
RMSE for midwest (n=626): 0.7984
RMSE for west (n=1030): 0.4169
RMSE for northeast (n=309): 0.8125
Final score: 0.7327164273457496


0.7327164273457496

Test run of crossval to check scorer

In [86]:
xgb_cv = cross_validate(xgb_model, 
                        X_train, 
                        y_train_adj, 
                        scoring=regional_scorer, 
                        return_train_score=True)

In [87]:
print(f"Train: {xgb_cv['train_score'].mean()}")
print(f"Test: {xgb_cv['test_score'].mean()}")
xgb_cv

Train: -0.27170464845212283
Test: -0.7686739521485018


{'fit_time': array([2.08083367, 2.20314503, 2.09651113, 2.06565738, 2.1649003 ]),
 'score_time': array([0.33353615, 0.33875418, 0.28201723, 0.33447814, 0.33862782]),
 'test_score': array([-0.77387006, -0.76368498, -0.73119046, -0.79161926, -0.783005  ]),
 'train_score': array([-0.26383289, -0.28239701, -0.27817458, -0.27092972, -0.26318904])}

Updating the scorecard.

In [94]:
scores_df = update_scorecard(xgb_model, 'XGBoost_Base', scores_df)
scores_df

RMSE for south (n=2744): 0.9031
RMSE for midwest (n=626): 0.7984
RMSE for west (n=1030): 0.4169
RMSE for northeast (n=309): 0.8125
Final score: 0.7327164273457496


Model Name  CV_Train   CV_Test         Difference  Unseen Data
0         dummy_model  1.658196  1.657582  Underfit: (0.001)     1.653806
1         LogReg_Base  1.658196  1.657582  Underfit: (0.001)     1.653806
2          LogReg_GS1  0.983089  0.984196   Overfit: (0.001)     0.952627
3          LogReg_GS2  0.978392  0.982059   Overfit: (0.004)     0.951622
4          LogReg_GS3  0.974411  0.981109   Overfit: (0.007)     0.952596
5  Random_Forest_Base  0.004530  0.836388   Overfit: (0.832)     0.773322
6   Random_Forest_GS1  0.052565  0.831655   Overfit: (0.779)     0.788095
7   Random_Forest_GS2  0.076151  0.814660   Overfit: (0.739)     0.775325
8   Random_Forest_GS3  0.067371  0.808106   Overfit: (0.741)     0.783564
9        XGBoost_Base  0.271705  0.768674   Overfit: (0.497)     0.732716

### XGBoost Gridsearch

XGBoost sees some moderate improvement on the cross validation score when compared to the baseline run with scores going from .768 to .764.

In [95]:
xgb_params = {
    'xgb__learning_rate':[.1, .05, .2, .3],
    'xgb__n_estimators':[100, 200, 300, 400]
}
xgb_gs = GridSearchCV(estimator=xgb_model, 
                      param_grid=xgb_params, 
                      scoring=regional_scorer, 
                      cv=5, 
                      verbose=1)

xgb_gs.fit(X_train, y_train_adj);

Fitting 5 folds for each of 16 candidates, totalling 80 fits


In [96]:
xgb_gs.best_params_

{'xgb__learning_rate': 0.2, 'xgb__n_estimators': 200}

In [97]:
xgb_gs.best_score_

-0.7645919304785711

Updating the scorecard.

In [98]:
scores_df = update_scorecard(xgb_gs.best_estimator_, 'XGBoost_GS', scores_df)
scores_df

RMSE for south (n=2744): 0.894
RMSE for midwest (n=626): 0.823
RMSE for west (n=1030): 0.3978
RMSE for northeast (n=309): 0.78
Final score: 0.7236974579802316


Model Name  CV_Train   CV_Test         Difference  Unseen Data
0          dummy_model  1.658196  1.657582  Underfit: (0.001)     1.653806
1          LogReg_Base  1.658196  1.657582  Underfit: (0.001)     1.653806
2           LogReg_GS1  0.983089  0.984196   Overfit: (0.001)     0.952627
3           LogReg_GS2  0.978392  0.982059   Overfit: (0.004)     0.951622
4           LogReg_GS3  0.974411  0.981109   Overfit: (0.007)     0.952596
5   Random_Forest_Base  0.004530  0.836388   Overfit: (0.832)     0.773322
6    Random_Forest_GS1  0.052565  0.831655   Overfit: (0.779)     0.788095
7    Random_Forest_GS2  0.076151  0.814660   Overfit: (0.739)     0.775325
8    Random_Forest_GS3  0.067371  0.808106   Overfit: (0.741)     0.783564
9         XGBoost_Base  0.271705  0.768674   Overfit: (0.497)     0.732716
10          XGBoost_GS  0.211160  0.764592   Overfit: (0.553)     0.723697

### Best XGBoost Model

The XGBoost final model is scoring better on cross validated data and on the unseen validation data with the validated data improving from .732 to .723. That being said, the model is even more over fit on training data than when just using the baseline XGB model with cross validated training results coming in at .211.

In [99]:
best_xgb_model = Pipeline([
    ('ColTrans', ColTrans),
    ('xgb', XGBClassifier(learning_rate= 0.2,
                          n_estimators = 200,
                          random_state=42))
])

In [100]:
best_xgb_cv = cross_validate(best_xgb_model, 
                             X_train, 
                             y_train_adj, 
                             scoring=regional_scorer, 
                             return_train_score=True)

In [101]:
print(f"Train: {best_xgb_cv['train_score'].mean()}")
print(f"Test: {best_xgb_cv['test_score'].mean()}")
best_xgb_cv

Train: -0.21116028380884452
Test: -0.7645919304785711


{'fit_time': array([2.92352986, 3.06136012, 2.93630719, 2.90041828, 3.0217483 ]),
 'score_time': array([0.33873558, 0.33577967, 0.28651166, 0.34249973, 0.34550548]),
 'test_score': array([-0.80036305, -0.75295327, -0.71766184, -0.77744225, -0.77453924]),
 'train_score': array([-0.20585199, -0.20828675, -0.22562312, -0.21283745, -0.2032021 ])}

**Testing on Validation Data**

In [102]:
best_xgb_model.fit(X_train, y_train_adj);

In [103]:
best_xgb_gs_val_pred = best_xgb_model.predict(X_val)
best_xgb_gs_val_pred = best_xgb_gs_val_pred+1

In [104]:
check_score(val_score, best_xgb_gs_val_pred)

RMSE for south (n=2744): 0.894
RMSE for midwest (n=626): 0.823
RMSE for west (n=1030): 0.3978
RMSE for northeast (n=309): 0.78
Final score: 0.7236974579802316


0.7236974579802316

## LinearSVC Model

LinearSVC does not perform well with this dataset and scored 1.48 on cross validated test data. I will attempt a GridSearch to improve results.

In [105]:
svc_model = Pipeline([
    ('ColTrans', ColTrans),
    ('svc', LinearSVC(random_state=42))
])

In [106]:
svc_model.fit(X_train, y_train);

In [107]:
svc_model.score(X_val, y_val)

0.4332129963898917

In [108]:
svc_pred = svc_model.predict(X_val)

In [109]:
check_score(val_score, svc_pred)

RMSE for south (n=2744): 0.9751
RMSE for midwest (n=626): 1.5982
RMSE for west (n=1030): 2.8527
RMSE for northeast (n=309): 1.1892
Final score: 1.6538061708774303


1.6538061708774303

In [110]:
svc_cv = cross_validate(svc_model, 
                        X_train, 
                        y_train, 
                        scoring=regional_scorer, 
                        return_train_score=True)

In [111]:
print(f"Train: {svc_cv['train_score'].mean()}")
print(f"Test: {svc_cv['test_score'].mean()}")
svc_cv

Train: -1.4762110688311139
Test: -1.4867126514015816


{'fit_time': array([4.27243924, 4.43056631, 4.35035849, 4.34207129, 4.41135097]),
 'score_time': array([0.30850101, 0.30359387, 0.24200082, 0.30803251, 0.31235695]),
 'test_score': array([-1.37439214, -1.98060022, -1.20816215, -1.65495332, -1.21545543]),
 'train_score': array([-1.37929896, -1.94271165, -1.20098151, -1.65907389, -1.19898933])}

Updating the scorecard.

In [112]:
scores_df = update_scorecard(svc_model, 'SVC_Base', scores_df)
scores_df

RMSE for south (n=2744): 0.9751
RMSE for midwest (n=626): 1.5982
RMSE for west (n=1030): 2.8527
RMSE for northeast (n=309): 1.1892
Final score: 1.6538061708774303


Model Name  CV_Train   CV_Test         Difference  Unseen Data
0          dummy_model  1.658196  1.657582  Underfit: (0.001)     1.653806
1          LogReg_Base  1.658196  1.657582  Underfit: (0.001)     1.653806
2           LogReg_GS1  0.983089  0.984196   Overfit: (0.001)     0.952627
3           LogReg_GS2  0.978392  0.982059   Overfit: (0.004)     0.951622
4           LogReg_GS3  0.974411  0.981109   Overfit: (0.007)     0.952596
5   Random_Forest_Base  0.004530  0.836388   Overfit: (0.832)     0.773322
6    Random_Forest_GS1  0.052565  0.831655   Overfit: (0.779)     0.788095
7    Random_Forest_GS2  0.076151  0.814660   Overfit: (0.739)     0.775325
8    Random_Forest_GS3  0.067371  0.808106   Overfit: (0.741)     0.783564
9         XGBoost_Base  0.271705  0.768674   Overfit: (0.497)     0.732716
10          XGBoost_GS  0.211160  0.764592   Overfit: (0.553)     0.723697
11            SVC_Base  1.476211  1.486713   Overfit: (0.011)     1.653806

### SVC GridSearch

LinearSVC GridSearch did not improve results. I will not be attempting any further modeling or making a best SVC Model.

In [113]:
svc_params = {
    'svc__fit_intercept':[True, False],
    'svc__class_weight':[None, 'balanced']
}
svc_gs = GridSearchCV(estimator=svc_model, 
                      param_grid=svc_params, 
                      scoring=regional_scorer, 
                      cv=5, 
                      verbose=1,
                      n_jobs=2)

svc_gs.fit(X_train, y_train);

Fitting 5 folds for each of 4 candidates, totalling 20 fits


In [114]:
svc_gs.best_params_

{'svc__class_weight': None, 'svc__fit_intercept': True}

In [115]:
svc_gs.best_score_

-1.4867126514015816

Updating the scorecard.

In [116]:
scores_df = update_scorecard(svc_gs.best_estimator_, 'SVC_GS', scores_df)
scores_df

RMSE for south (n=2744): 0.9751
RMSE for midwest (n=626): 1.5982
RMSE for west (n=1030): 2.8527
RMSE for northeast (n=309): 1.1892
Final score: 1.6538061708774303


Model Name  CV_Train   CV_Test         Difference  Unseen Data
0          dummy_model  1.658196  1.657582  Underfit: (0.001)     1.653806
1          LogReg_Base  1.658196  1.657582  Underfit: (0.001)     1.653806
2           LogReg_GS1  0.983089  0.984196   Overfit: (0.001)     0.952627
3           LogReg_GS2  0.978392  0.982059   Overfit: (0.004)     0.951622
4           LogReg_GS3  0.974411  0.981109   Overfit: (0.007)     0.952596
5   Random_Forest_Base  0.004530  0.836388   Overfit: (0.832)     0.773322
6    Random_Forest_GS1  0.052565  0.831655   Overfit: (0.779)     0.788095
7    Random_Forest_GS2  0.076151  0.814660   Overfit: (0.739)     0.775325
8    Random_Forest_GS3  0.067371  0.808106   Overfit: (0.741)     0.783564
9         XGBoost_Base  0.271705  0.768674   Overfit: (0.497)     0.732716
10          XGBoost_GS  0.211160  0.764592   Overfit: (0.553)     0.723697
11            SVC_Base  1.476211  1.486713   Overfit: (0.011)     1.653806
12              SVC_GS  1.476211  1.486713   Overfit: (0.011)     1.653806

## MLP Model

MLP performs about as well as the Dummy Model on both test and training data. I will attempt to GridSearch the model to improve results.

In [117]:
mlp_model = Pipeline([
    ('ColTrans', ColTrans),
    ('mlp', MLPClassifier(early_stopping=True, random_state=42))
])

In [118]:
mlp_model.fit(X_train, y_train);

In [119]:
mlp_model.score(X_val, y_val)

0.4332129963898917

In [120]:
mlp_val_pred = mlp_model.predict(X_val)

In [121]:
check_score(val_score, mlp_val_pred)

RMSE for south (n=2744): 0.9751
RMSE for midwest (n=626): 1.5982
RMSE for west (n=1030): 2.8527
RMSE for northeast (n=309): 1.1892
Final score: 1.6538061708774303


1.6538061708774303

In [122]:
mlp_cv = cross_validate(mlp_model, 
                        X_train, 
                        y_train, 
                        scoring=regional_scorer, 
                        return_train_score=True)

In [123]:
print(f"Train: {mlp_cv['train_score'].mean()}")
print(f"Test: {mlp_cv['test_score'].mean()}")
mlp_cv

Train: -1.658196064281532
Test: -1.6575819051725245


{'fit_time': array([1.43138242, 1.53153372, 1.40632749, 1.40784168, 1.49684024]),
 'score_time': array([0.30635405, 0.30799866, 0.25181723, 0.31250501, 0.31000733]),
 'test_score': array([-1.64969162, -1.62809574, -1.66571971, -1.65495332, -1.68944913]),
 'train_score': array([-1.65997003, -1.66581971, -1.65633035, -1.65907389, -1.64978634])}

Updating the scorecard.

In [124]:
scores_df = update_scorecard(mlp_model, 'MLP_Base', scores_df)
scores_df

RMSE for south (n=2744): 0.9751
RMSE for midwest (n=626): 1.5982
RMSE for west (n=1030): 2.8527
RMSE for northeast (n=309): 1.1892
Final score: 1.6538061708774303


Model Name  CV_Train   CV_Test         Difference  Unseen Data
0          dummy_model  1.658196  1.657582  Underfit: (0.001)     1.653806
1          LogReg_Base  1.658196  1.657582  Underfit: (0.001)     1.653806
2           LogReg_GS1  0.983089  0.984196   Overfit: (0.001)     0.952627
3           LogReg_GS2  0.978392  0.982059   Overfit: (0.004)     0.951622
4           LogReg_GS3  0.974411  0.981109   Overfit: (0.007)     0.952596
5   Random_Forest_Base  0.004530  0.836388   Overfit: (0.832)     0.773322
6    Random_Forest_GS1  0.052565  0.831655   Overfit: (0.779)     0.788095
7    Random_Forest_GS2  0.076151  0.814660   Overfit: (0.739)     0.775325
8    Random_Forest_GS3  0.067371  0.808106   Overfit: (0.741)     0.783564
9         XGBoost_Base  0.271705  0.768674   Overfit: (0.497)     0.732716
10          XGBoost_GS  0.211160  0.764592   Overfit: (0.553)     0.723697
11            SVC_Base  1.476211  1.486713   Overfit: (0.011)     1.653806
12              SVC_GS  1.476211  1.486713   Overfit: (0.011)     1.653806
13            MLP_Base  1.658196  1.657582  Underfit: (0.001)     1.653806

### First MLPClassifier Gridsearch

The first MLPClassifier GridSearch improved cross validated results from 1.657 to 1.293. I will attempt a second GridSearch to further improve the results.

In [125]:
mlp_params = {
    'mlp__learning_rate':['constant', 'invscaling', 'adaptive'],
    'mlp__activation':['relu', 'tanh'],
    'mlp__hidden_layer_sizes':[(100,), 
                               (100, 40, 50, 10), 
                               (10, 40, 50), 
                               (40, 40, 40, 23)]
}
mlp_gs = GridSearchCV(estimator=mlp_model, 
                      param_grid=mlp_params, 
                      scoring=regional_scorer, 
                      cv=5, 
                      verbose=1, 
                      n_jobs=2)

mlp_gs.fit(X_train, y_train);

Fitting 5 folds for each of 24 candidates, totalling 120 fits


In [126]:
mlp_gs.best_params_

{'mlp__activation': 'relu',
 'mlp__hidden_layer_sizes': (100, 40, 50, 10),
 'mlp__learning_rate': 'constant'}

In [127]:
mlp_gs.best_score_

-1.2935554220077328

Updating the scorecard.

In [128]:
scores_df = update_scorecard(mlp_gs.best_estimator_, 'MLP_GS1', scores_df)
scores_df

RMSE for south (n=2744): 0.9751
RMSE for midwest (n=626): 1.5982
RMSE for west (n=1030): 2.8527
RMSE for northeast (n=309): 1.1892
Final score: 1.6538061708774303


Model Name  CV_Train   CV_Test         Difference  Unseen Data
0          dummy_model  1.658196  1.657582  Underfit: (0.001)     1.653806
1          LogReg_Base  1.658196  1.657582  Underfit: (0.001)     1.653806
2           LogReg_GS1  0.983089  0.984196   Overfit: (0.001)     0.952627
3           LogReg_GS2  0.978392  0.982059   Overfit: (0.004)     0.951622
4           LogReg_GS3  0.974411  0.981109   Overfit: (0.007)     0.952596
5   Random_Forest_Base  0.004530  0.836388   Overfit: (0.832)     0.773322
6    Random_Forest_GS1  0.052565  0.831655   Overfit: (0.779)     0.788095
7    Random_Forest_GS2  0.076151  0.814660   Overfit: (0.739)     0.775325
8    Random_Forest_GS3  0.067371  0.808106   Overfit: (0.741)     0.783564
9         XGBoost_Base  0.271705  0.768674   Overfit: (0.497)     0.732716
10          XGBoost_GS  0.211160  0.764592   Overfit: (0.553)     0.723697
11            SVC_Base  1.476211  1.486713   Overfit: (0.011)     1.653806
12              SVC_GS  1.476211  1.486713   Overfit: (0.011)     1.653806
13            MLP_Base  1.658196  1.657582  Underfit: (0.001)     1.653806
14             MLP_GS1  1.293554  1.293555    Good Fit: (0.0)     1.653806

### Second MLPClassifier GridSearch

The second iteration of GridSearch for MLPClassifier did not perform any better than the first. I will not pursue this method any further.

In [129]:
mlp_params2 = {
    'mlp__learning_rate':['constant'],
    'mlp__activation':['relu', 'tanh'],
    'mlp__hidden_layer_sizes':[(100, 100, 100, 100, 100), 
                               (100, 40, 50, 10), 
                               (10, 40, 50), 
                               (40, 40, 40, 23)],
    'mlp__max_iter':[200, 300, 400]
}
mlp_gs2 = GridSearchCV(estimator=mlp_model, 
                      param_grid=mlp_params2, 
                      scoring=regional_scorer, 
                      cv=5, 
                      verbose=1, 
                      n_jobs=4)

mlp_gs2.fit(X_train, y_train);

Fitting 5 folds for each of 24 candidates, totalling 120 fits


In [130]:
mlp_gs2.best_params_

{'mlp__activation': 'relu',
 'mlp__hidden_layer_sizes': (100, 40, 50, 10),
 'mlp__learning_rate': 'constant',
 'mlp__max_iter': 200}

In [131]:
mlp_gs2.best_score_

-1.2935554220077328

Updating the scorecard.

In [132]:
scores_df = update_scorecard(mlp_gs2.best_estimator_, 'MLP_GS2', scores_df)
scores_df

RMSE for south (n=2744): 0.9751
RMSE for midwest (n=626): 1.5982
RMSE for west (n=1030): 2.8527
RMSE for northeast (n=309): 1.1892
Final score: 1.6538061708774303


Model Name  CV_Train   CV_Test         Difference  Unseen Data
0          dummy_model  1.658196  1.657582  Underfit: (0.001)     1.653806
1          LogReg_Base  1.658196  1.657582  Underfit: (0.001)     1.653806
2           LogReg_GS1  0.983089  0.984196   Overfit: (0.001)     0.952627
3           LogReg_GS2  0.978392  0.982059   Overfit: (0.004)     0.951622
4           LogReg_GS3  0.974411  0.981109   Overfit: (0.007)     0.952596
5   Random_Forest_Base  0.004530  0.836388   Overfit: (0.832)     0.773322
6    Random_Forest_GS1  0.052565  0.831655   Overfit: (0.779)     0.788095
7    Random_Forest_GS2  0.076151  0.814660   Overfit: (0.739)     0.775325
8    Random_Forest_GS3  0.067371  0.808106   Overfit: (0.741)     0.783564
9         XGBoost_Base  0.271705  0.768674   Overfit: (0.497)     0.732716
10          XGBoost_GS  0.211160  0.764592   Overfit: (0.553)     0.723697
11            SVC_Base  1.476211  1.486713   Overfit: (0.011)     1.653806
12              SVC_GS  1.476211  1.486713   Overfit: (0.011)     1.653806
13            MLP_Base  1.658196  1.657582  Underfit: (0.001)     1.653806
14             MLP_GS1  1.293554  1.293555    Good Fit: (0.0)     1.653806
15             MLP_GS2  1.293554  1.293555    Good Fit: (0.0)     1.653806

## K-Nearest Neighbors Model

K-Nearest Neighbors performs moderately well compared to the other algorithms. The Regional RMSE is 1.07 on unseen validation data and the cross validated score is 1.11. The model does show signs of overfitting with a cross validated score of .911 on training data that I will attempt to fix during a GridSearch.

In [133]:
knn = Pipeline([
    ('ColTrans', ColTrans),
    ('knn', KNeighborsClassifier())
])

In [134]:
knn.fit(X_train, y_train);

In [135]:
knn.score(X_val, y_val)

0.5973667445317478

In [136]:
knn_pred = knn.predict(X_val)

In [137]:
check_score(val_score, knn_pred)

RMSE for south (n=2744): 0.9383
RMSE for midwest (n=626): 1.1769
RMSE for west (n=1030): 0.8752
RMSE for northeast (n=309): 1.2897
Final score: 1.0700377479276688


1.0700377479276688

In [138]:
knn_cv = cross_validate(knn, 
                        X_train, 
                        y_train, 
                        scoring=regional_scorer, 
                        return_train_score=True)

In [139]:
print(f"Train: {knn_cv['train_score'].mean()}")
print(f"Test: {knn_cv['test_score'].mean()}")
knn_cv

Train: -0.9116659571264305
Test: -1.1144825396071767


{'fit_time': array([1.20681143, 1.29611993, 1.19859672, 1.18440056, 1.25252557]),
 'score_time': array([0.34150815, 0.35749865, 0.30168819, 0.35400367, 0.34751439]),
 'test_score': array([-1.1854625 , -1.07529758, -1.07921362, -1.10068514, -1.13175386]),
 'train_score': array([-0.912322  , -0.9139052 , -0.90338932, -0.92285927, -0.90585399])}

Updating the scorecard.

In [140]:
scores_df = update_scorecard(knn, 'KNN_Base', scores_df)
scores_df

RMSE for south (n=2744): 0.9383
RMSE for midwest (n=626): 1.1769
RMSE for west (n=1030): 0.8752
RMSE for northeast (n=309): 1.2897
Final score: 1.0700377479276688


Model Name  CV_Train   CV_Test         Difference  Unseen Data
0          dummy_model  1.658196  1.657582  Underfit: (0.001)     1.653806
1          LogReg_Base  1.658196  1.657582  Underfit: (0.001)     1.653806
2           LogReg_GS1  0.983089  0.984196   Overfit: (0.001)     0.952627
3           LogReg_GS2  0.978392  0.982059   Overfit: (0.004)     0.951622
4           LogReg_GS3  0.974411  0.981109   Overfit: (0.007)     0.952596
5   Random_Forest_Base  0.004530  0.836388   Overfit: (0.832)     0.773322
6    Random_Forest_GS1  0.052565  0.831655   Overfit: (0.779)     0.788095
7    Random_Forest_GS2  0.076151  0.814660   Overfit: (0.739)     0.775325
8    Random_Forest_GS3  0.067371  0.808106   Overfit: (0.741)     0.783564
9         XGBoost_Base  0.271705  0.768674   Overfit: (0.497)     0.732716
10          XGBoost_GS  0.211160  0.764592   Overfit: (0.553)     0.723697
11            SVC_Base  1.476211  1.486713   Overfit: (0.011)     1.653806
12              SVC_GS  1.476211  1.486713   Overfit: (0.011)     1.653806
13            MLP_Base  1.658196  1.657582  Underfit: (0.001)     1.653806
14             MLP_GS1  1.293554  1.293555    Good Fit: (0.0)     1.653806
15             MLP_GS2  1.293554  1.293555    Good Fit: (0.0)     1.653806
16            KNN_Base  0.911666  1.114483   Overfit: (0.203)     1.070038

### First K-Nearest Neighbors Gridsearch

The first iteration of GridSearch for KNN shows improved results from a cross validated score of 1.11 to .997. I will look to improve these results in a second iteration of GridSearch.

In [141]:
knn_params = {
    'knn__weights':['uniform', 'distance'],
    'knn__p':[1, 2, 3, 4]
}
knn_gs = GridSearchCV(estimator=knn, 
                      param_grid=knn_params, 
                      scoring=regional_scorer, 
                      cv=5, 
                      verbose=1, 
                      n_jobs=2)

knn_gs.fit(X_train, y_train);

Fitting 5 folds for each of 8 candidates, totalling 40 fits


In [142]:
knn_gs.best_params_

{'knn__p': 1, 'knn__weights': 'distance'}

In [143]:
knn_gs.best_score_

-0.9972729582651667

Updating the scorecard.

In [144]:
scores_df = update_scorecard(knn_gs.best_estimator_, 'KNN_GS1', scores_df)
scores_df

RMSE for south (n=2744): 0.9524
RMSE for midwest (n=626): 1.0491
RMSE for west (n=1030): 0.6812
RMSE for northeast (n=309): 1.0582
Final score: 0.9352315744769308


Model Name  CV_Train   CV_Test         Difference  Unseen Data
0          dummy_model  1.658196  1.657582  Underfit: (0.001)     1.653806
1          LogReg_Base  1.658196  1.657582  Underfit: (0.001)     1.653806
2           LogReg_GS1  0.983089  0.984196   Overfit: (0.001)     0.952627
3           LogReg_GS2  0.978392  0.982059   Overfit: (0.004)     0.951622
4           LogReg_GS3  0.974411  0.981109   Overfit: (0.007)     0.952596
5   Random_Forest_Base  0.004530  0.836388   Overfit: (0.832)     0.773322
6    Random_Forest_GS1  0.052565  0.831655   Overfit: (0.779)     0.788095
7    Random_Forest_GS2  0.076151  0.814660   Overfit: (0.739)     0.775325
8    Random_Forest_GS3  0.067371  0.808106   Overfit: (0.741)     0.783564
9         XGBoost_Base  0.271705  0.768674   Overfit: (0.497)     0.732716
10          XGBoost_GS  0.211160  0.764592   Overfit: (0.553)     0.723697
11            SVC_Base  1.476211  1.486713   Overfit: (0.011)     1.653806
12              SVC_GS  1.476211  1.486713   Overfit: (0.011)     1.653806
13            MLP_Base  1.658196  1.657582  Underfit: (0.001)     1.653806
14             MLP_GS1  1.293554  1.293555    Good Fit: (0.0)     1.653806
15             MLP_GS2  1.293554  1.293555    Good Fit: (0.0)     1.653806
16            KNN_Base  0.911666  1.114483   Overfit: (0.203)     1.070038
17             KNN_GS1  0.000000  0.997273   Overfit: (0.997)     0.935232

### Second K-Nearest Neighbors GridSearch

The second iteration of GridSearch shows improved results from a cross validated score of .997 to a cross validated score of .975.

In [145]:
knn_params2 = {
    'knn__n_neighbors':[1, 2, 3, 4, 5],
    'knn__weights':['distance'],
    'knn__p':[1, 2, 3, 4]
}
knn_gs2 = GridSearchCV(estimator=knn, 
                      param_grid=knn_params2, 
                      scoring=regional_scorer, 
                      cv=5, 
                      verbose=1, 
                      n_jobs=2)

knn_gs2.fit(X_train, y_train);

Fitting 5 folds for each of 20 candidates, totalling 100 fits


In [146]:
knn_gs2.best_params_

{'knn__n_neighbors': 3, 'knn__p': 1, 'knn__weights': 'distance'}

In [147]:
knn_gs2.best_score_

-0.9755125348123379

Updating the scorecard.

In [148]:
scores_df = update_scorecard(knn_gs2.best_estimator_, 'KNN_GS2', scores_df)
scores_df

RMSE for south (n=2744): 0.9562
RMSE for midwest (n=626): 1.0291
RMSE for west (n=1030): 0.6726
RMSE for northeast (n=309): 0.9102
Final score: 0.8920461891255786


Model Name  CV_Train   CV_Test         Difference  Unseen Data
0          dummy_model  1.658196  1.657582  Underfit: (0.001)     1.653806
1          LogReg_Base  1.658196  1.657582  Underfit: (0.001)     1.653806
2           LogReg_GS1  0.983089  0.984196   Overfit: (0.001)     0.952627
3           LogReg_GS2  0.978392  0.982059   Overfit: (0.004)     0.951622
4           LogReg_GS3  0.974411  0.981109   Overfit: (0.007)     0.952596
5   Random_Forest_Base  0.004530  0.836388   Overfit: (0.832)     0.773322
6    Random_Forest_GS1  0.052565  0.831655   Overfit: (0.779)     0.788095
7    Random_Forest_GS2  0.076151  0.814660   Overfit: (0.739)     0.775325
8    Random_Forest_GS3  0.067371  0.808106   Overfit: (0.741)     0.783564
9         XGBoost_Base  0.271705  0.768674   Overfit: (0.497)     0.732716
10          XGBoost_GS  0.211160  0.764592   Overfit: (0.553)     0.723697
11            SVC_Base  1.476211  1.486713   Overfit: (0.011)     1.653806
12              SVC_GS  1.476211  1.486713   Overfit: (0.011)     1.653806
13            MLP_Base  1.658196  1.657582  Underfit: (0.001)     1.653806
14             MLP_GS1  1.293554  1.293555    Good Fit: (0.0)     1.653806
15             MLP_GS2  1.293554  1.293555    Good Fit: (0.0)     1.653806
16            KNN_Base  0.911666  1.114483   Overfit: (0.203)     1.070038
17             KNN_GS1  0.000000  0.997273   Overfit: (0.997)     0.935232
18             KNN_GS2  0.000000  0.975513   Overfit: (0.976)     0.892046

### K-Nearest Neighbors Best Model

The KNN best model is performing ok with a cross validated score of .975 and a score of .892 on unseen validation data. The model is incredibly overfit with a perfect training score.

In [149]:
best_knn = Pipeline([
    ('ColTrans', ColTrans),
    ('knn', KNeighborsClassifier(n_neighbors=3,
                                p=1,
                                weights='distance'))
])

In [150]:
best_knn_cv = cross_validate(best_knn, 
                             X_train, 
                             y_train, 
                             scoring=regional_scorer, 
                             return_train_score=True)

In [151]:
print(f"Train: {best_knn_cv['train_score'].mean()}")
print(f"Test: {best_knn_cv['test_score'].mean()}")
best_knn_cv

Train: 0.0
Test: -0.9755125348123379


{'fit_time': array([1.21180344, 1.3399291 , 1.21552873, 1.20750952, 1.29003668]),
 'score_time': array([0.36102509, 0.36466885, 0.3110137 , 0.37173057, 0.3752048 ]),
 'test_score': array([-1.01255625, -0.93576172, -0.94862547, -0.99144258, -0.98917666]),
 'train_score': array([-0., -0., -0., -0., -0.])}

**Testing on Validation Data**

In [152]:
best_knn.fit(X_train, y_train);

In [153]:
best_knn_pred = best_knn.predict(X_val)

In [154]:
check_score(val_score, best_knn_pred)

RMSE for south (n=2744): 0.9562
RMSE for midwest (n=626): 1.0291
RMSE for west (n=1030): 0.6726
RMSE for northeast (n=309): 0.9102
Final score: 0.8920461891255786


0.8920461891255786

## Stacking Classifier Model

I'm setting up a stacking classifier with the best models and their hyperparameters to see how they perform in tandem. The stacking classifier received a score of .719 on unseen validation data. It also has cross validated results of .788 on test data and .392 on training data.

I am only including models that received a score of less than 1.00 on cross validated training data. I am excluding KNN due to the extreme overfitting of the model.

While this model has a cross validated test score that is lower than the XGBoost model, it is much less overfit and performs better on unseen validation data.

In [155]:
estimators = [
    ('logreg', LogisticRegression(random_state=42,
                                 fit_intercept=False,
                                 max_iter=175,
                                 penalty=None,
                                 solver='newton-cg')),
    ('rf', RandomForestClassifier(criterion='entropy',
                                  min_samples_split=3,
                                  n_estimators=150,
                                  random_state=42)),
    ('xgb', XGBClassifier(learning_rate= 0.2,
                          n_estimators = 200,
                          random_state=42))
]

In [156]:
stack_model = Pipeline([
    ('ColTrans', ColTrans),
    ('stack', StackingClassifier(estimators))
])

In [157]:
stack_model.fit(X_train, y_train_adj);

In [158]:
stack_model.score(X_val, y_val_adj)

0.6742408154597579

In [159]:
stack_val_pred = stack_model.predict(X_val)
stack_val_pred = stack_val_pred+1

In [160]:
check_score(val_score, stack_val_pred)

RMSE for south (n=2744): 0.9001
RMSE for midwest (n=626): 0.8043
RMSE for west (n=1030): 0.3791
RMSE for northeast (n=309): 0.7964
Final score: 0.7199775137157071


0.7199775137157071

In [161]:
stack_cv = cross_validate(stack_model, 
                          X_train, 
                          y_train_adj, 
                          scoring=regional_scorer, 
                          return_train_score=True)

In [162]:
print(f"Train: {stack_cv['train_score'].mean()}")
print(f"Test: {stack_cv['test_score'].mean()}")
stack_cv

Train: -0.3922973330342381
Test: -0.7887982868606599


{'fit_time': array([48.09307528, 46.60119295, 47.2515974 , 46.85318065, 47.15291429]),
 'score_time': array([0.37225437, 0.37751412, 0.31352592, 0.38240147, 0.37601829]),
 'test_score': array([-0.80409216, -0.76702976, -0.74844025, -0.81701968, -0.80740958]),
 'train_score': array([-0.40927534, -0.37963539, -0.38918324, -0.39265756, -0.39073513])}

Updating the scorecard.

In [163]:
scores_df = update_scorecard(stack_model, 'Stack1', scores_df)
scores_df

RMSE for south (n=2744): 0.9001
RMSE for midwest (n=626): 0.8043
RMSE for west (n=1030): 0.3791
RMSE for northeast (n=309): 0.7964
Final score: 0.7199775137157071


Model Name  CV_Train   CV_Test         Difference  Unseen Data
0          dummy_model  1.658196  1.657582  Underfit: (0.001)     1.653806
1          LogReg_Base  1.658196  1.657582  Underfit: (0.001)     1.653806
2           LogReg_GS1  0.983089  0.984196   Overfit: (0.001)     0.952627
3           LogReg_GS2  0.978392  0.982059   Overfit: (0.004)     0.951622
4           LogReg_GS3  0.974411  0.981109   Overfit: (0.007)     0.952596
5   Random_Forest_Base  0.004530  0.836388   Overfit: (0.832)     0.773322
6    Random_Forest_GS1  0.052565  0.831655   Overfit: (0.779)     0.788095
7    Random_Forest_GS2  0.076151  0.814660   Overfit: (0.739)     0.775325
8    Random_Forest_GS3  0.067371  0.808106   Overfit: (0.741)     0.783564
9         XGBoost_Base  0.271705  0.768674   Overfit: (0.497)     0.732716
10          XGBoost_GS  0.211160  0.764592   Overfit: (0.553)     0.723697
11            SVC_Base  1.476211  1.486713   Overfit: (0.011)     1.653806
12              SVC_GS  1.476211  1.486713   Overfit: (0.011)     1.653806
13            MLP_Base  1.658196  1.657582  Underfit: (0.001)     1.653806
14             MLP_GS1  1.293554  1.293555    Good Fit: (0.0)     1.653806
15             MLP_GS2  1.293554  1.293555    Good Fit: (0.0)     1.653806
16            KNN_Base  0.911666  1.114483   Overfit: (0.203)     1.070038
17             KNN_GS1  0.000000  0.997273   Overfit: (0.997)     0.935232
18             KNN_GS2  0.000000  0.975513   Overfit: (0.976)     0.892046
19              Stack1  0.392297  0.788798   Overfit: (0.397)     0.719978

### Second Stacking Classifier Model

In this iteration of the model, I am not using the "best" XGBoost model and instead will be using the untuned model to compare. I am making this change because the two models have very similar results on cross validated test data, while the untuned model is much less overfit.

While this model performed worse on the unseen validation data. It has a higher cross validated test score (.779) and is less overfit (.415) on the training data.

In [164]:
estimators2 = [
    ('logreg', LogisticRegression(random_state=42,
                                 fit_intercept=False,
                                 max_iter=175,
                                 penalty=None,
                                 solver='newton-cg')),
    ('rf', RandomForestClassifier(criterion='entropy',
                                  min_samples_split=3,
                                  n_estimators=150,
                                  random_state=42)),
    ('xgb', XGBClassifier(random_state=42))
]

In [165]:
stack_model2 = Pipeline([
    ('ColTrans', ColTrans),
    ('stack2', StackingClassifier(estimators2))
])

In [166]:
stack_model2.fit(X_train, y_train_adj);

In [167]:
stack_model2.score(X_val, y_val_adj)

0.6750902527075813

In [168]:
stack_val_pred2 = stack_model2.predict(X_val)
stack_val_pred2 = stack_val_pred2+1

In [169]:
check_score(val_score, stack_val_pred2)

RMSE for south (n=2744): 0.8956
RMSE for midwest (n=626): 0.8152
RMSE for west (n=1030): 0.4051
RMSE for northeast (n=309): 0.8341
Final score: 0.7375015647214597


0.7375015647214597

In [170]:
stack_cv2 = cross_validate(stack_model2, 
                          X_train, 
                          y_train_adj, 
                          scoring=regional_scorer, 
                          return_train_score=True)

In [171]:
print(f"Train: {stack_cv2['train_score'].mean()}")
print(f"Test: {stack_cv2['test_score'].mean()}")
stack_cv2

Train: -0.415859998890553
Test: -0.7793028863598923


{'fit_time': array([44.21412706, 42.63399458, 43.35133576, 42.65951347, 43.240978  ]),
 'score_time': array([0.36850834, 0.38551521, 0.31171989, 0.3757174 , 0.38901663]),
 'test_score': array([-0.78652868, -0.75813897, -0.76706417, -0.80204771, -0.78273492]),
 'train_score': array([-0.4189947 , -0.40629408, -0.41070106, -0.4155496 , -0.42776056])}

Updating the scorecard.

In [172]:
scores_df = update_scorecard(stack_model2, 'Stack2', scores_df)
scores_df

RMSE for south (n=2744): 0.8956
RMSE for midwest (n=626): 0.8152
RMSE for west (n=1030): 0.4051
RMSE for northeast (n=309): 0.8341
Final score: 0.7375015647214597


Model Name  CV_Train   CV_Test         Difference  Unseen Data
0          dummy_model  1.658196  1.657582  Underfit: (0.001)     1.653806
1          LogReg_Base  1.658196  1.657582  Underfit: (0.001)     1.653806
2           LogReg_GS1  0.983089  0.984196   Overfit: (0.001)     0.952627
3           LogReg_GS2  0.978392  0.982059   Overfit: (0.004)     0.951622
4           LogReg_GS3  0.974411  0.981109   Overfit: (0.007)     0.952596
5   Random_Forest_Base  0.004530  0.836388   Overfit: (0.832)     0.773322
6    Random_Forest_GS1  0.052565  0.831655   Overfit: (0.779)     0.788095
7    Random_Forest_GS2  0.076151  0.814660   Overfit: (0.739)     0.775325
8    Random_Forest_GS3  0.067371  0.808106   Overfit: (0.741)     0.783564
9         XGBoost_Base  0.271705  0.768674   Overfit: (0.497)     0.732716
10          XGBoost_GS  0.211160  0.764592   Overfit: (0.553)     0.723697
11            SVC_Base  1.476211  1.486713   Overfit: (0.011)     1.653806
12              SVC_GS  1.476211  1.486713   Overfit: (0.011)     1.653806
13            MLP_Base  1.658196  1.657582  Underfit: (0.001)     1.653806
14             MLP_GS1  1.293554  1.293555    Good Fit: (0.0)     1.653806
15             MLP_GS2  1.293554  1.293555    Good Fit: (0.0)     1.653806
16            KNN_Base  0.911666  1.114483   Overfit: (0.203)     1.070038
17             KNN_GS1  0.000000  0.997273   Overfit: (0.997)     0.935232
18             KNN_GS2  0.000000  0.975513   Overfit: (0.976)     0.892046
19              Stack1  0.392297  0.788798   Overfit: (0.397)     0.719978
20              Stack2  0.415860  0.779303   Overfit: (0.363)     0.737502

### Third Iteration of Stacking Classifier

Since the Random Forest Model scored .067 on cross validated training data and .808 on cross validated test data, I will be taking it out of this version of the stacking classifier to try and further reduce overfitting.

This model is less overfit than the previous iteration, however it performs worse on unseen validation data and worse on cross validated test data.

In [173]:
estimators3 = [
    ('logreg', LogisticRegression(random_state=42,
                                 fit_intercept=False,
                                 max_iter=175,
                                 penalty=None,
                                 solver='newton-cg')),
    ('xgb', XGBClassifier(random_state=42))
]

In [174]:
stack_model3 = Pipeline([
    ('ColTrans', ColTrans),
    ('stack3', StackingClassifier(estimators3))
])

In [175]:
stack_model3.fit(X_train, y_train_adj);

In [176]:
stack_model3.score(X_val, y_val_adj)

0.6719048630282438

In [177]:
stack_val_pred3 = stack_model3.predict(X_val)
stack_val_pred3 = stack_val_pred3+1

In [178]:
check_score(val_score, stack_val_pred3)

RMSE for south (n=2744): 0.9059
RMSE for midwest (n=626): 0.8336
RMSE for west (n=1030): 0.4063
RMSE for northeast (n=309): 0.8419
Final score: 0.7469129213950463


0.7469129213950463

In [179]:
stack_cv3 = cross_validate(stack_model3, 
                          X_train, 
                          y_train_adj, 
                          scoring=regional_scorer, 
                          return_train_score=True)

In [180]:
print(f"Train: {stack_cv3['train_score'].mean()}")
print(f"Test: {stack_cv3['test_score'].mean()}")
stack_cv3

Train: -0.4292451797483495
Test: -0.788109006249498


{'fit_time': array([26.2828455 , 24.8337059 , 25.21585536, 24.75838065, 25.05565047]),
 'score_time': array([0.30653286, 0.31974506, 0.27471161, 0.32365441, 0.31772304]),
 'test_score': array([-0.79848001, -0.76733538, -0.77473691, -0.81545725, -0.78453549]),
 'train_score': array([-0.4207207 , -0.41142169, -0.43287377, -0.42746691, -0.45374282])}

Updating the scorecard.

In [181]:
scores_df = update_scorecard(stack_model3, 'Stack3', scores_df)
scores_df

RMSE for south (n=2744): 0.9059
RMSE for midwest (n=626): 0.8336
RMSE for west (n=1030): 0.4063
RMSE for northeast (n=309): 0.8419
Final score: 0.7469129213950463


Model Name  CV_Train   CV_Test         Difference  Unseen Data
0          dummy_model  1.658196  1.657582  Underfit: (0.001)     1.653806
1          LogReg_Base  1.658196  1.657582  Underfit: (0.001)     1.653806
2           LogReg_GS1  0.983089  0.984196   Overfit: (0.001)     0.952627
3           LogReg_GS2  0.978392  0.982059   Overfit: (0.004)     0.951622
4           LogReg_GS3  0.974411  0.981109   Overfit: (0.007)     0.952596
5   Random_Forest_Base  0.004530  0.836388   Overfit: (0.832)     0.773322
6    Random_Forest_GS1  0.052565  0.831655   Overfit: (0.779)     0.788095
7    Random_Forest_GS2  0.076151  0.814660   Overfit: (0.739)     0.775325
8    Random_Forest_GS3  0.067371  0.808106   Overfit: (0.741)     0.783564
9         XGBoost_Base  0.271705  0.768674   Overfit: (0.497)     0.732716
10          XGBoost_GS  0.211160  0.764592   Overfit: (0.553)     0.723697
11            SVC_Base  1.476211  1.486713   Overfit: (0.011)     1.653806
12              SVC_GS  1.476211  1.486713   Overfit: (0.011)     1.653806
13            MLP_Base  1.658196  1.657582  Underfit: (0.001)     1.653806
14             MLP_GS1  1.293554  1.293555    Good Fit: (0.0)     1.653806
15             MLP_GS2  1.293554  1.293555    Good Fit: (0.0)     1.653806
16            KNN_Base  0.911666  1.114483   Overfit: (0.203)     1.070038
17             KNN_GS1  0.000000  0.997273   Overfit: (0.997)     0.935232
18             KNN_GS2  0.000000  0.975513   Overfit: (0.976)     0.892046
19              Stack1  0.392297  0.788798   Overfit: (0.397)     0.719978
20              Stack2  0.415860  0.779303   Overfit: (0.363)     0.737502
21              Stack3  0.429245  0.788109   Overfit: (0.359)     0.746913

### Final Iteration of Stacking Classifier

While the untuned version of K Nearest Neighbors only scored 1.11 on cross validated test data, the model was only slightly overfit. I will be adding it back into the stacking classifier to see if it is able to help with performance and overfitting.

This model performs the best of all models, scoring .738 on unseen validation data and returning cross validated results of .776 for test data and .413 on training data. While it does not have the lowest validation score of all data, it does have the best combined results of all three indicators.

In [182]:
estimators4 = [
    ('logreg', LogisticRegression(random_state=42,
                                 fit_intercept=False,
                                 max_iter=175,
                                 penalty=None,
                                 solver='newton-cg')),
    ('knn', KNeighborsClassifier(weights='distance')),
    ('rf', RandomForestClassifier(criterion='entropy',
                                  min_samples_split=3,
                                  n_estimators=150,
                                  random_state=42)),
    ('xgb', XGBClassifier(random_state=42))
]

In [183]:
stack_model4 = Pipeline([
    ('ColTrans', ColTrans),
    ('stack4', StackingClassifier(estimators4))
])

In [184]:
stack_model4.fit(X_train, y_train_adj);

In [185]:
stack_model4.score(X_val, y_val_adj)

0.675302612019537

In [186]:
stack_val_pred4 = stack_model4.predict(X_val)
stack_val_pred4 = stack_val_pred4+1

In [187]:
check_score(val_score, stack_val_pred4)

RMSE for south (n=2744): 0.8968
RMSE for midwest (n=626): 0.824
RMSE for west (n=1030): 0.399
RMSE for northeast (n=309): 0.8341
Final score: 0.7384900178900713


0.7384900178900713

In [188]:
stack_cv4 = cross_validate(stack_model4, 
                          X_train, 
                          y_train_adj, 
                          scoring=regional_scorer, 
                          return_train_score=True)

In [189]:
print(f"Train: {stack_cv4['train_score'].mean()}")
print(f"Test: {stack_cv4['test_score'].mean()}")
stack_cv4

Train: -0.4135086642180109
Test: -0.7768198987829074


{'fit_time': array([43.77444267, 42.32908344, 42.6063621 , 42.15849686, 42.48474336]),
 'score_time': array([0.39528155, 0.40203714, 0.34105277, 0.40929842, 0.40003824]),
 'test_score': array([-0.77884424, -0.75703243, -0.75939241, -0.80114741, -0.787683  ]),
 'train_score': array([-0.41473293, -0.40478804, -0.40929326, -0.41881375, -0.41991534])}

Updating the scorecard.

In [190]:
scores_df = update_scorecard(stack_model4, 'Stack4', scores_df)
scores_df

RMSE for south (n=2744): 0.9001
RMSE for midwest (n=626): 0.8043
RMSE for west (n=1030): 0.3791
RMSE for northeast (n=309): 0.7964
Final score: 0.7199775137157071


Model Name  CV_Train   CV_Test         Difference  Unseen Data
0          dummy_model  1.658196  1.657582  Underfit: (0.001)     1.653806
1          LogReg_Base  1.658196  1.657582  Underfit: (0.001)     1.653806
2           LogReg_GS1  0.983089  0.984196   Overfit: (0.001)     0.952627
3           LogReg_GS2  0.978392  0.982059   Overfit: (0.004)     0.951622
4           LogReg_GS3  0.974411  0.981109   Overfit: (0.007)     0.952596
5   Random_Forest_Base  0.004530  0.836388   Overfit: (0.832)     0.773322
6    Random_Forest_GS1  0.052565  0.831655   Overfit: (0.779)     0.788095
7    Random_Forest_GS2  0.076151  0.814660   Overfit: (0.739)     0.775325
8    Random_Forest_GS3  0.067371  0.808106   Overfit: (0.741)     0.783564
9         XGBoost_Base  0.271705  0.768674   Overfit: (0.497)     0.732716
10          XGBoost_GS  0.211160  0.764592   Overfit: (0.553)     0.723697
11            SVC_Base  1.476211  1.486713   Overfit: (0.011)     1.653806
12              SVC_GS  1.476211  1.486713   Overfit: (0.011)     1.653806
13            MLP_Base  1.658196  1.657582  Underfit: (0.001)     1.653806
14             MLP_GS1  1.293554  1.293555    Good Fit: (0.0)     1.653806
15             MLP_GS2  1.293554  1.293555    Good Fit: (0.0)     1.653806
16            KNN_Base  0.911666  1.114483   Overfit: (0.203)     1.070038
17             KNN_GS1  0.000000  0.997273   Overfit: (0.997)     0.935232
18             KNN_GS2  0.000000  0.975513   Overfit: (0.976)     0.892046
19              Stack1  0.392297  0.788798   Overfit: (0.397)     0.719978
20              Stack2  0.415860  0.779303   Overfit: (0.363)     0.737502
21              Stack3  0.429245  0.788109   Overfit: (0.359)     0.746913
22              Stack1  0.392297  0.788798   Overfit: (0.397)     0.719978

# Best Model

In [191]:
scores_df

Model Name  CV_Train   CV_Test         Difference  Unseen Data
0          dummy_model  1.658196  1.657582  Underfit: (0.001)     1.653806
1          LogReg_Base  1.658196  1.657582  Underfit: (0.001)     1.653806
2           LogReg_GS1  0.983089  0.984196   Overfit: (0.001)     0.952627
3           LogReg_GS2  0.978392  0.982059   Overfit: (0.004)     0.951622
4           LogReg_GS3  0.974411  0.981109   Overfit: (0.007)     0.952596
5   Random_Forest_Base  0.004530  0.836388   Overfit: (0.832)     0.773322
6    Random_Forest_GS1  0.052565  0.831655   Overfit: (0.779)     0.788095
7    Random_Forest_GS2  0.076151  0.814660   Overfit: (0.739)     0.775325
8    Random_Forest_GS3  0.067371  0.808106   Overfit: (0.741)     0.783564
9         XGBoost_Base  0.271705  0.768674   Overfit: (0.497)     0.732716
10          XGBoost_GS  0.211160  0.764592   Overfit: (0.553)     0.723697
11            SVC_Base  1.476211  1.486713   Overfit: (0.011)     1.653806
12              SVC_GS  1.476211  1.486713   Overfit: (0.011)     1.653806
13            MLP_Base  1.658196  1.657582  Underfit: (0.001)     1.653806
14             MLP_GS1  1.293554  1.293555    Good Fit: (0.0)     1.653806
15             MLP_GS2  1.293554  1.293555    Good Fit: (0.0)     1.653806
16            KNN_Base  0.911666  1.114483   Overfit: (0.203)     1.070038
17             KNN_GS1  0.000000  0.997273   Overfit: (0.997)     0.935232
18             KNN_GS2  0.000000  0.975513   Overfit: (0.976)     0.892046
19              Stack1  0.392297  0.788798   Overfit: (0.397)     0.719978
20              Stack2  0.415860  0.779303   Overfit: (0.363)     0.737502
21              Stack3  0.429245  0.788109   Overfit: (0.359)     0.746913
22              Stack1  0.392297  0.788798   Overfit: (0.397)     0.719978

In [192]:
stop

NameError: name 'stop' is not defined

### Visualizing Errors

My final model has a cross validated Regional RMSE of .7768 on validation data. My model correctly predicted the right severity level over 2/3rds of the time. On sample sites that were incorrectly predicted, 71% were only off by one point. 

In [ ]:
val_score['Error'] = abs(val_score['severity'] - val_score['pred'])

In [ ]:
val_score.head()

In [ ]:
final_val_df = sat_train[['uid', 'geometry']]

In [ ]:
plotting_final_df = pd.merge(left=val_score, right=final_val_df, how='left', on='uid')
plotting_final_df.head()

In [ ]:
final_error_df = plotting_final_df[plotting_final_df['Error'] != 0]

Was 100% correct on about 2/3rds of the datapoints.

In [ ]:
1-(len(final_error_df)/len(plotting_final_df))

71% of the missed datapoints were only off by one point. 

In [ ]:
final_error_df['Error'].value_counts(normalize=True)

In [ ]:
fig, ax = plt.subplots(figsize=(20,15))
# Setting a basemap of the USA
USA_map = world[world['name'] == 'United States of America'].plot(ax=ax, 
                                                                  color='#C2C2C2', 
                                                                  edgecolor='black')

# Labeling the recorded severity levels
gpd.GeoDataFrame(final_error_df).plot(ax=ax,
                    column='Error',
                    categorical=True,
                    cmap='plasma_r',
                    alpha= .6,
                    markersize=60,
                    legend=True,
                    legend_kwds={'markerscale':1.25,
                                'fontsize':'large',
                                'title':'Error of Prediction',
                                'title_fontsize':'x-large',
                                'ncols':2}
                      )


ax.set_xlim([-128, -65])
ax.set_ylim([23, 53])
ax.set_axis_off()
ax.set_title('Model Prediction Errors', fontsize=25);

We can see that the vast majority of errors came from the southern region. 

In [ ]:
final_error_df['region'].value_counts()

In [ ]:
final_error_df.groupby('Error')['region'].value_counts()

# Conclusions

## Analysis

This analysis has run dozens of models to predict the severity of harmful algal blooms in inland bodies of water. Ultimately the best model was the Stacking Classifier using Logistic Regression, K-Nearest Neighbors, Random Forest, and XGBoost. The 5-fold cross validated score was .7768 for the Regional Root Mean Squared Error. When Tested on unseen validation data, the model scored:

- RMSE for south (n=2744): 0.9059
- RMSE for midwest (n=626): 0.8336
- RMSE for west (n=1030): 0.4063
- RMSE for northeast (n=309): 0.8419
- Final score: 0.7469129213950463

The model has a much better predictive ability for the western region than the other regions. This makes sense when we look at the EPA data and can visually see that the western region has a relatively consistent severity average of between about 3.5 and 4. Visually, the Northeast and Midwest appear to have the most variability in their severity, though the model is better able to predict for these regions than for the south.

It is important to note that the safety of people, animals, and the environment is at stake, so the focus of this analysis was to minimize the error in the predictions rather than to achieve a high level of accuracy. The use of root mean squared error was appropriate for this purpose as it measures the deviation between the predicted values and the actual values, and the goal was to keep this deviation as low as possible.

This analysis has proven that machine learning using existing EPA sample data, NASA Satellite Imagery, and NOAA climate data can be used to better predict the severity of harmful algal blooms in inland bodies of water.

Additional Funding to continue research and implement improvements on the model is vital to the safety and security of our aquatic environments. In the future considerations section below, I highlight some of the areas where additional research could benefit the model.

## Future Considerations

There are several areas for improvement that could further enhance the performance of the predictive model. These include:

- Incorporating multiple satellite images for each site: Utilizing multiple satellite images for each site has the potential to significantly improve the model's predictive ability by providing more data for the model to learn from.


- Using more climate data: In this analysis, only two points of climate data (temperature from 1 and 5 days prior to EPA sampling) were used. Incorporating more climate data points, such as wind gusts and precipitation, from the NOAA HRRR system would greatly improve the model's ability to predict HABs.


- Using more temperature data: In this analysis, only one temperature point was taken for each day. Utilizing hourly temperature data from the NOAA HRRR system would significantly improve the model's ability to predict HABs.

Additional funding and resources for continued research and implementation of these improvements are crucial for ensuring the safety and protection of the environment and the communities that rely on freshwater sources.